### PR TITLE
Dropped `age` column from patient registration and annotated age in patient viewset

### PR DESCRIPTION
**Pull request by rithviknishad**
## Proposed Changes

- Dropped DB field `age` from Patient Registration
- Adds field `death_datetime` in Patient Registration
- Custom Migration to copy last_consultation's death_datetime to patient registration if present
- Updated consultation save method to update patient registration's death_datetime too if consultation has death_datetime.
- Annotated patient viewset with age to calculate age based on data of birth/year of birth up to date of death or now.

### Associated Issue

- https://github.com/coronasafe/care_fe/issues/7383
- FE: https://github.com/coronasafe/care_fe/pull/7411

> [!NOTE] 
> ### Custom Migrations
> 
> - Populates `PatientRegistration.death_datetime` with `instance.last_consultation.death_datetime` if present.

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

File: care/facility/api/serializers/patient.py [hunk 1]
```diff
@@ -252,11 +258,15 @@ def validate(self, attrs):
         validated = super().validate(attrs)
         if (
             not self.partial
-            and not validated.get("age")
+            and not validated.get("year_of_birth")
             and not validated.get("date_of_birth")
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion
            and (not validated.get("year_of_birth")
            or not validated.get("date_of_birth"))
```

File: care/facility/models/patient_icmr.py [hunk 2]
```diff
@@ -45,26 +46,18 @@ class Meta:
     #         instance.__class__ = PatientSampleICMR
     #     return instance
 
+    def get_age_delta(self):
+        start = self.date_of_birth or timezone.datetime(self.year_of_birth, 1, 1).date()
+        end = (self.death_datetime or timezone.now()).date()
+        return relativedelta(end - start)
```
Comment (sainak) [inline care/facility/models/patient_icmr.py]: dateutil.relativedelta package is only used here, can we remove it?
we could just convert dates to datetime with h:mm as 0:00
and return the resulting timedelta

File: care/facility/models/patient_icmr.py [hunk 3]
```diff
@@ -45,26 +46,18 @@ class Meta:
     #         instance.__class__ = PatientSampleICMR
     #     return instance
 
+    def get_age_delta(self):
```
Comment (sainak) [inline care/facility/models/patient_icmr.py:49]:
```suggestion
    def get_age_delta(self) -> timedelta:
```

File: care/facility/models/patient_icmr.py [hunk 4]
```diff
@@ -45,26 +46,18 @@ class Meta:
...
+        start = self.date_of_birth or timezone.datetime(self.year_of_birth, 1, 1).date()
+        end = (self.death_datetime or timezone.now()).date()
+        return relativedelta(end - start)
+
     @property
     def age_years(self):
```
Comment (sainak) [inline care/facility/models/patient_icmr.py]:
```suggestion
    def age_years(self) -> int:
```

File: care/facility/models/patient_icmr.py [hunk 5]
```diff
@@ -45,26 +46,18 @@ class Meta:
...
-        else:
-            age_years = relativedelta(
-                datetime.datetime.now(),
-                datetime.datetime(year=self.year_of_birth, month=1, day=1),
-            ).years
-        return age_years
+        return self.get_age_delta().years
 
     @property
     def age_months(self):
```
Comment (sainak) [inline care/facility/models/patient_icmr.py]:
```suggestion
    def age_months(self) -> int:
```

File: care/facility/migrations/0422_remove_patientregistration_age_and_add_patientregistration_death_datetime.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("facility", "0421_merge_20240318_1434"),
+    ]
+
+    def populate_patientregistration_death_datetime(apps, schema_editor):
+        PatientRegistration = apps.get_model("facility", "PatientRegistration")
+
+        patients = PatientRegistration.objects.filter(
```
Comment (sainak) [inline care/facility/migrations/0422_remove_patientregistration_age_and_add_patientregistration_death_datetime.py]:
```suggestion
        patients = PatientRegistration.objects.only("last_consultation").filter(
```

**Review by sainak — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Comment by nihal467** [comment]
LGTM

**Review by Ashesh3 — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**